Move the Maybe Datacenter into the ConsulClient, drop from requests

ketzacoatl opened the pull request:
This update move the Datacenter into the ConsulClient data structure. In effect, we're saying a "client connection" specifies the datacenter. So if you want to interact with multiple datacenters in a consul cluster, you would have multiple clients. I think this is a good thing, as MOST of the time, there's a single datacenter we're targeting, and if/when there are multiple, having separate clients is a great way to keep the code clean and obvious.

With this change, the datacenter doesn't need to be passed around outside the client, we can simply rely on the client data structure to contain the info we need for the interaction with consul.

With this update, users will need to change their calling code to use the library (dropping the Maybe Datacenter from most functions, and specifying it when initializing the client). An example of that update looks like:

```
+dc1 :: Datacenter
+dc1 = Datacenter "dc1"
+
 newClient :: IO ConsulClient
-newClient = initializeConsulClient "localhost" consulPort emptyHttpManager
+newClient = initializeConsulClient "localhost" consulPort dc1 emptyHttpManager
```

```
 testGetInvalidKey :: TestTree
 testGetInvalidKey = testCase "testGetInvalidKey" $ do
   client@ConsulClient{..} <- newClient
-  x <- getKey client "nokey" Nothing Nothing Nothing
+  x <- getKey client "nokey" Nothing Nothing
   assertEqual "testGetInvalidKey: Found a key that doesn't exist" x Nothing
```

While we will want to update `async` to `withAsync` as well, these changes will result in a version bump to `0.6.0`.

This contributes to #32.

ketzacoatl commented:
I forgot to mention, tests are still passing:

![image](https://user-images.githubusercontent.com/10122937/101781363-b4ba2b00-3ac5-11eb-93b2-97c10b91d50a.png)

ketzacoatl commented:
Here is the failure from Travis:

![image](https://user-images.githubusercontent.com/10122937/101814015-88190a00-3aeb-11eb-83d7-01b1e363b0fd.png)

I'm not sure if that is a transient error or if there's an update to be made to the API call.

ketzacoatl commented:
With consul `1.9.0`, two (session) tests fail:

![image](https://user-images.githubusercontent.com/10122937/101844000-5fa80480-3b19-11eb-977e-4d0852526fe1.png)

That should probably be fixed separately.

ketzacoatl commented:
All those issues noted above have been resolved (see #36 and #38), and this PR has been updated/rebased. I also have a version of this PR that is based on #39, though I'm not sure that one is good to merge, so I haven't pushed it. @nh2, LMK if I should update this after merging that PR, or if this should be first.

ketzacoatl commented:
Tests are passing:

![image](https://user-images.githubusercontent.com/10122937/102014178-b7658c00-3d22-11eb-81a5-306a434b0dc4.png)

nh2 commented on src/Network/Consul/Types.hs:
```diff
@@ -40,12 +40,13 @@ import Data.Word
+  , ccHostname :: Text
+  , ccPort :: PortNumber
+  , ccWithTls :: Bool
+  , ccDatacenter :: Datacenter
```
I believe it would be better to have this be `Maybe Datacenter`, because you can leave it away entirely in consul requests (and I think in most of my consul usage I do leave it away).

I also think we do not need to update the signatures of `initializeConsulClient` and `initializeTlsConsulClient` in this PR: We can default the datacenter to `Nothing` there, and users can use record updates if they need to change the datacenter (see main review comment for an example).

ketzacoatl replied:
Thanks for the feedback, I can make each of those updates!

ketzacoatl replied:
I've updated this in the latest version of the PR.

ketzacoatl replied:
I've applied these updates in the latest PR.

nh2 commented on tests/Main.hs:
```diff
@@ -36,8 +36,11 @@ import SocketUtils (isPortOpen, simpleSockAddr)
 consulPort = 18500
 
+dc1 :: Datacenter
+dc1 = Datacenter "dc1"
```
Why does `"dc1"` work out of the box, is that Consul's default datacenter name?

ketzacoatl replied:
Yep, you can see it in the output from consul when we run tests.

ketzacoatl replied:
I've applied these updates in the latest PR.

nh2 reviewed (changes requested):
Your PR description is great, we should copy it into the changelog when we make this API backwards incompatble release.

The only thing you might want to add is that it is still very easy to provide a custom datacenter by updating the `client` record:

```diff
-  x <- getKey client "nokey" Nothing Nothing Nothing
+  x <- getKey client{ ccDatacenter = Datacenter "mydc" } "nokey" Nothing Nothing
```

I've made a small change request to maintain the `Maybe`-ness of the `ccDatacenter`.

ketzacoatl commented:
> Your PR description is great, we should copy it into the changelog when we make this API backwards incompatble release.
> 
> The only thing you might want to add is that it is still very easy to provide a custom datacenter by updating the `client` record:
> 
> ```diff
> -  x <- getKey client "nokey" Nothing Nothing Nothing
> +  x <- getKey client{ ccDatacenter = Datacenter "mydc" } "nokey" Nothing Nothing
> ```
> 
> I've made a small change request to maintain the `Maybe`-ness of the `ccDatacenter`.

This should now be complete.

ketzacoatl commented:
OK, #39 was updated and merged, so I've rebased this PR and confirmed tests are passing. This also addresses the `Maybe Datacenter` request, so I believe this is ready for re-review and merge.

ketzacoatl commented:
@nh2 maybe you can review? I am blocked without a thumbs up or more changes to make.

ketzacoatl commented:
After fixing CI for #50, I bumped the version, added a CHANGELOG, updated the tests to work with #48, and made sure CI is still looking good. I believe this is now ready to merge.
